db/firebase/firebaseUtil.js

```diff
@@ -0,0 +1,404 @@
+        let userRef = getFactotumDoc().collection('guilds').doc(guildId).collection('members').where('discordId', '==', id).limit(1);
+        let user = (await userRef.get()).docs[0];
+
+        if (user) {
+            /** @type {FirebaseUser} */
+            let data = user.data();
+
+            data.types.forEach((value, index, array) => {
+                if (value.isVerified) {
+                    value.isAttending = true;
+                    value.AttendingTimestamp = admin.firestore.Timestamp.now();
+                }
+            });
+
+            user.ref.update(data);
+        } else {
+            throw new Error('The discord id provided was not found!');
+        }
+    },
+
+    /**
+     * check if codex is set to active (very hacky atm, it's just a document in the "codex" collection with a boolean 
+     * field called "active")
+     * @param {String} guildId 
+     * @returns boolean of whether codex is set to active
+     */
+    async checkCodexActive(guildId) {
+        let ref = getFactotumDoc().collection('guilds').doc(guildId).collection('codex').doc('active');
+        let activeRef = await ref.get();
+        const data = activeRef.data();
+        return data.active;
+    },
+
+    /**
+     * stores email to firebase collection
+     * @param {String} guildId 
+     * @param {String} collection - name of collection to store email in
+     * @param {String} email - user's email
+     */
+    async saveToFirebase(guildId, collection, email) {
+        let ref = getFactotumDoc().collection('guilds').doc(guildId).collection(collection).doc(email.toLowerCase());
+        /** @type {FirebaseUser} */
+        let data = {
+            email: email.toLowerCase()
+        };
+
+        await ref.set(data);
+    },
+
+    async lookupById(guildId, memberId) {
+        const userRef = getFactotumDoc().collection('guilds').doc(guildId).collection('members').where('discordId', '==', memberId).limit(1);
+        const user = (await userRef.get()).docs[0];
+        return user ? user.data().email : undefined;
+    },
+
+    async saveToLeaderboard(guildId, memberId) {
+        const userRef = getFactotumDoc().collection('guilds').doc(guildId).collection('questionsLeaderboard').doc(memberId);
+        const user = await userRef.get();
+        if (user.exists) {
+            const data = user.data();
+            data.points++;
+            await userRef.update(data);
+        } else {
+            const data = { memberId, points: 1 };
+            await userRef.set(data);
+        }
+    },
+
+    async retrieveLeaderboard(guildId) {
+        const snapshot = (await getFactotumDoc().collection('guilds').doc(guildId).collection('questionsLeaderboard').get()).docs;
+        const winners = snapshot.map(doc => doc.data()).sort((a, b) => b.points - a.points);
+        return winners;
+    },
+
+};
+
+function getFactotumDoc() {
+    return apps.get('nwPlusBotAdmin').firestore().collection('ExternalProjects').doc('Factotum');
```

Comment:
Should we use the _db object or directly reference with apps.get?

Reply:
yep just addressed @daniel-panhead ^^